server/aesdsocket.c

```diff
@@ -359,171 +396,370 @@ int run_daemon(void)
+        {
+            syslog(LOG_ERR, "Thread creation for new client failed");
+            free(freshNode);
+            return ERROR;
+        }
 
-    syslog(LOG_INFO, "Accepted connection from %s", ip_address);
+        // Insert node into list
+        SLIST_INSERT_HEAD(&head, freshNode, nodes);
+        freshNode = NULL;
 
-    return 0;
+        // Check for thread completion and join them
+        SLIST_FOREACH(freshNode, &head, nodes)
+        {
+            if (freshNode->thread_data.isThreadComplete)
+            {
+                if (pthread_join(freshNode->thread_data.threadId, &threadRetVal) == ERROR)
+                {
+                    syslog(LOG_ERR, "Failed to join completed thread");
+                    return ERROR;
+                }
+                if (threadRetVal == NULL)
+                {
+                    return ERROR;
+                }
+                syslog(LOG_INFO, "Successfully joined thread %ld", freshNode->thread_data.threadId);
+            }
+        }
+    }
+
+    return SUCCESS;
 }
 
 
 /**
- * @brief Receive data over a connection and append it to a file.
+ * @brief Function to handle both receiving and sending data through a client socket.
  * 
- * This function receives data from a client connected to the socket and appends it to the file located at /var/tmp/aesdsocketdata.
- *
- * @return 0 on successful operation, -1 otherwise
+ * This function does the following:
+ * 1. Receives data from the client and writes it to the file /var/tmp/aesdsocketdata
+ * 2. Reads the content from the same file and sends it back to the client.
+ * 
+ * @param thread_param Pointer to the thread data structure
+ * @return Returns the pointer to the thread data structure
  */
-int receive_and_store_data(void)
+void* client_data_handler(void *thread_param)
 {
-    ssize_t bytes_received = 1;  // Number of bytes received (initialize to non-zero to start loop)
-    char buf[BUF_LEN];  // Buffer to temporarily hold received data
-    int ret_status;  // Status returned by the write operation
+    int result;
+    // variables for receiving data
+    ssize_t bytes_received = 0;
+    char receive_buffer[BUF_LEN];
+
+    // variables for sending data
+    ssize_t bytes_sent = 0;
+    char send_buffer[BUF_LEN];
+    ssize_t bytes_from_file = 1;
+
+    memset(receive_buffer, 0, BUF_LEN);
+    memset(send_buffer, 0, BUF_LEN);
+
+    ClientThreadData_t *thread_data_ptr = (ClientThreadData_t*)thread_param;
+
+    inet_ntop(thread_data_ptr->pClientAddr->ss_family,
+              get_in_addr((struct sockaddr *)&(thread_data_ptr->pClientAddr)),
+              s, sizeof s);
+    syslog(LOG_INFO, "New connection established: %s", s);
+
+    syslog(LOG_INFO, "Thread %ld initialized", thread_data_ptr->threadId);
 
-    // Loop for receiving and storing data
-    while (bytes_received > 0)
+    // Lock mutex to protect file
+    result = pthread_mutex_lock(thread_data_ptr->pMutex);
```

Comment:
@Suraj-Ajjampur is this the minimum locking interval?  What is `pMutex` locking?

Reply:
The updated function aims to minimize the duration for which the mutex is locked by unlocking it immediately after each write and send operation. This allows for better concurrency but may result in indeterministic writes and reads if multiple threads interact with the same file. Additionally, there are missing unlocks in the error paths, which could lead to a deadlock.

```
void* client_data_handler(void *thread_param)
{
    int result;
    // variables for receiving data
    ssize_t bytes_received = 0;
    char receive_buffer[BUF_LEN];

    // variables for sending data
    ssize_t bytes_sent = 0;
    char send_buffer[BUF_LEN];
    ssize_t bytes_from_file = 1;

    memset(receive_buffer, 0, BUF_LEN);
    memset(send_buffer, 0, BUF_LEN);

    ClientThreadData_t *thread_data_ptr = (ClientThreadData_t*)thread_param;

    inet_ntop(thread_data_ptr->pClientAddr->ss_family,
              get_in_addr((struct sockaddr *)&(thread_data_ptr->pClientAddr)),
              s, sizeof s);
    syslog(LOG_INFO, "New connection established: %s", s);

    syslog(LOG_INFO, "Thread %ld initialized", thread_data_ptr->threadId);

    // Lock mutex to protect file
    result = pthread_mutex_lock(thread_data_ptr->pMutex);
    if (result == ERROR)
    {
        syslog(LOG_ERR, "Failed to acquire mutex");
        return NULL;
    }

    // Initialize the condition variable
    void *newline_found = NULL;

    // Execute loop as long as the condition variable is NULL
    while (newline_found == NULL)
    {
        bytes_received = recv(thread_data_ptr->clientSocketFd, receive_buffer, BUF_LEN, 0);
        if (bytes_received == ERROR)
        {
            syslog(LOG_ERR, "Data reception unsuccessful");
            return NULL;
        }

        result = write(dataFileDescriptor, receive_buffer, bytes_received);
        if (result == ERROR)
        {
            syslog(LOG_ERR, "Unsuccessful file write operation");
            return NULL;
        }

        // Unlock mutex
        result = pthread_mutex_unlock(thread_data_ptr->pMutex);
        if (result == ERROR)
        {
        syslog(LOG_ERR, "Failed to release mutex");
        return NULL;
        }

        // Update the condition variable
        newline_found = memchr(receive_buffer, '\n', bytes_received);
    }

    // Lock mutex to protect file
    result = pthread_mutex_lock(thread_data_ptr->pMutex);
    if (result == ERROR)
    {
        syslog(LOG_ERR, "Failed to acquire mutex");
        return NULL;
    }

    off_t seek_result = lseek(dataFileDescriptor, 0, SEEK_SET);
    if (seek_result == ERROR)
    {
        syslog(LOG_ERR, "Failed to seek in the file");
        return NULL;
    }

    // Loop as long as bytes_from_file is greater than 0
    while (bytes_from_file > 0)
    {
        bytes_from_file = read(dataFileDescriptor, send_buffer, BUF_LEN);
        if (bytes_from_file == ERROR)
        {
            syslog(LOG_ERR, "Failed to read file");
            return NULL;
        }

        // Check to avoid sending when read returns 0 (EOF)
        if (bytes_from_file > 0)
        {
            bytes_sent = send(thread_data_ptr->clientSocketFd, send_buffer, bytes_from_file, 0);
            if (bytes_sent == ERROR)
            {
                syslog(LOG_ERR, "Data transmission unsuccessful");
                return NULL;
            }

            // Unlock mutex
            result = pthread_mutex_unlock(thread_data_ptr->pMutex);
            if (result == ERROR)
            {
                syslog(LOG_ERR, "Failed to release mutex");
                return NULL;
            }
        }
    }

    close(thread_data_ptr->clientSocketFd);
    syslog(LOG_INFO, "Terminated connection: %s", s);

    thread_data_ptr->isThreadComplete = true;

    return thread_param;
}
```